lib.rs

```diff
@@ -24,6 +25,23 @@ mod token {
                 data: PSP22Data::new(supply, Self::env().caller()),
             }
         }
+
+        fn emit_events(&self, events: ink::prelude::vec::Vec<PSP22Event>) {
```

Comment:
sad thing is that this method has to be copied for every psp22 contract... maybe we could extract it from here and expose somehow for reuse?

if you go with the other suggestion, you might end up with a very generic, convenient method for emitting arbitrary event iterable from a contract, which sounds nice :)

Reply:
Unfortunately it's probably not possible to extract a method which uses `self.env()` outside the contract. Or maybe technically it's possible, but results in some templating nightmare.

This method is here temporary, after ink! allows defining events outside contracts, it will be gone and all `self.emit_events(events)` will be replaces with `for e in events { self.env().emit_event(e); }`

Reply:
I think you _could_ extract (or rather implement in the `PSP22Data`) something along these lines:
```rust
 fn emit_event<C: ContractBase, EE: EmitEvent<C>>(emitter: EE, event: Event) {
            emitter.emit_event(event);
        }
```